spinnaker/spinnaker_system/google_http_lb_upsert_test.py

```diff
@@ -0,0 +1,894 @@
+              'healthCheck': self.__proto_hc,
+            }
+          }
+        }
+      ],
+      'type': 'upsertLoadBalancer',
+      'availabilityZones': {bindings['TEST_GCE_REGION']: []},
+      'user': '[anonymous]'
+    }
+
+
+  def _make_bs_link(self, bs):
+    '''Make a fully-formatted backend service link.
+    '''
+    return (GCE_URL_PREFIX + self.bindings['GCE_PROJECT'] +
+            '/global/backendServices/' + bs)
+
+
+  def _make_hc_link(self, hc):
+    '''Make a fully-formatted health check link.
+    '''
+    return (GCE_URL_PREFIX + self.bindings['GCE_PROJECT'] +
+            '/global/httpHealthChecks/' + hc)
+
+
+  def _set_all_hcs(self, upsert, hc):
+    '''Set all health checks in upsert to hc.
+    '''
+    upsert['defaultService']['healthCheck'] = hc
+    for host_rule in upsert['hostRules']:
+      path_matcher = host_rule['pathMatcher']
+      path_matcher['defaultService']['healthCheck'] = hc
+      for path_rule in path_matcher['pathRules']:
+        path_rule['backendService']['healthCheck'] = hc
+
+
+  def _make_contract_clauses(self, contract_builder, upsert):
+    '''Add the proper predicates to the contract builder for a given
+    upsert description.
+    '''
+    host_rules = upsert['hostRules'] # Host rules will be distinct.
+    backend_services = [upsert['defaultService']]
+    for host_rule in host_rules:
+      path_matcher = host_rule['pathMatcher']
+      backend_services.append(path_matcher['defaultService'])
+      for path_rule in path_matcher['pathRules']:
+        backend_services.append(path_rule['backendService'])
+    health_checks = [service['healthCheck'] for service in backend_services]
+
+    hc_clause_builder = contract_builder.new_clause_builder('Health Checks Created',
+                                                            retryable_for_secs=30)
+    hc_clause_builder = hc_clause_builder.list_resource('httpHealthChecks')
+    for hc in health_checks:
+      hc_clause_builder.contains_pred_list(
+        [
+          jp.PathEqPredicate('name', hc['name']),
+          jp.PathEqPredicate('requestPath', hc['requestPath']),
+          jp.PathEqPredicate('port', hc['port'])
+        ]
+      )
+
+    bs_clause_builder = contract_builder.new_clause_builder('Backend Services Created',
+                                                            retryable_for_secs=30)
+    bs_clause_builder = bs_clause_builder.list_resource('backendServices')
+    for bs in backend_services:
+      bs_clause_builder.contains_pred_list(
+        [
+         jp.PathEqPredicate('name', bs['name']),
+         jp.PathEqPredicate('portName', 'http'),
+         jp.PathContainsPredicate('healthChecks[0]', bs['healthCheck']['name'])
+        ]
+      )
+
+    url_map_clause_builder = contract_builder.new_clause_builder('Url Map Created',
+                                                                 retryable_for_secs=30)
+    url_map_clause_builder = url_map_clause_builder.list_resource('urlMaps')
```

Comment:
I hate this github UI. Critique and Mondrian were so much better.
I'm not sure what I was seeing yesterday. I must not have seen the loop. You're right it doesnt make sense. In fact I think I missed that all these other loops are nested.

If I were doing a loop like this and couldnt write helper functions, i'd probably indent the block under #223 by 4 spaces rather than two (and add disable pylint indentation checking for that scope) to emphasize the scoping difference since it is so long.

But in this case what I'd do (but intentionally bit my tongue earlier) is create list of jp.AND predicate, build up that predicate with all the conditions, then add that predicate with build.add_constraint(host_result_predicate)

for hr in host_rules:
   pred_list = []
   pred_list.append(jp.AND(#226 - #228)
   pred_list.append(jp.AND( #244 for host in hr['hostPatterns']))
   ...
   builder.add_constraint(pred_list)

The above, I think, is clearer what you are saying. It adds another level of ANDs  ((a and b) AND (c and d)) rather than (a and b and c and d). I think it is easier to read and conceptualize, and might be easier to look at the reporting because you can see the context of the failure to the subexpression group rather than top-level element in a long chain.

Even without such a long nesting of four loops, when I thought it was just one, I'd still use the builder.add_constraint on a list. I think it makes the code cleaner since you can create the clause with the observation and predicates in a single literary statement.